build/index.js

```diff
@@ -261,7 +261,11 @@ async function main() {
     process.chdir(path.dirname(scriptPath));
 
     const eslintOptions = ["--ext", ".ts,.js,.tsx,.jsx", "."];
-    await npm("ci", "--no-update-notifier", "--no-fund");
+
+    if (command === "build" || command === "lint") {
```

Comment:
`test` also needs these packages (as of recently). At least I think so: If `test` doesn't run at all without calling `build` first, I agree with excluding it (even if it's a bit weird).

Maybe it makes more sense to just skip this step for certain tasks, e.g. `list` and `list-deps`? That way, the worst thing that could happen if we add a new command and forget to list it here is that the build takes longer than it'd have to. If we do it as an allowlist though, the worst thing that can happen if you add a new command is mysterious build errors on CI / other peoples machines.

Reply:
Yet another idea, you could complicate this logic a bit and introduce some `installRootModules` (or something) function that has a bit more logic:

1. It checks if `/node_modules` exists
2. It checks if the timestamp of `/node_modules/*` is older than that of `/package.json`
3. If that's the case, it runs the `npm ci` command - otherwise nothing

(While at it, it _really_ bothers me that we repeat that arcane `npm ci` command line now. Getting rid of the duplication could be adding a one liner like: `const npmCi = () => npm("ci", "--no-update-notifier", "--no-fund")` (no need for `async` since it already returns a promise and there's no need to call `await`)

Reply:
At the end of the day, we need to introduce state into the build. It's stateless right now, we're using the presence of node_modules to take care of state. If we do that, I'd like to do it in a less ad hoc way, using `npm` logic maybe. That should theoretically be possible... But we should also take care of not reinventing `gulp`, which we're *this* close to doing right now.

Reply:
The problem is that `npm` is missing that logic, calling `npm ci` takes forever because it verifies everything, I presume - it's simply not designed for monorepos containing multiple packages. Yarn 2.x is smarter there.

Anyway, I suppose in order to get this merged, it's better to not add that timestamp based logic, and just implement something like what I suggested in my first comment of this thread.